File: [O]math/[O]math/9-palindrome-number.py
class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x < 0:
            return False
        if x == 0:
            return True
        if x % 10 == 0:
            return False
        
        first_half = x
        second_half = 0
        while first_half > second_half:
            first_half, remainder = divmod(first_half, 10)
            second_half = second_half * 10 + remainder
            if first_half == second_half:
                return True
            if first_half == second_half // 10:
                return True
        return False

# time O(logn), base 10
# space O(1)
# using math

File: [O]math/[O]math/9-palindrome-number.java
class Solution {
    public boolean isPalindrome(int x) {
        if (x < 0) {
            return false;
        }
        if (x == 0) {
            return true;
        }
        if (x % 10 == 0) {
            return false;
        }
        int firstHalf = x;
        int secondHalf = 0;
        while (firstHalf > secondHalf) {
            int rem = firstHalf % 10;
            secondHalf = secondHalf * 10 + rem;
            firstHalf /= 10;
        }
        if (firstHalf < secondHalf) {
            secondHalf /= 10;
        }
        return firstHalf == secondHalf;
    }
}

// time O(logn), base 10
// space O(1)
// using math